casanovo/denovo/dataloaders.py

```diff
@@ -155,6 +170,35 @@ def _make_loader(
             shuffle=shuffle,
         )
 
+    def _make_db_loader(
+        self, dataset: torch.utils.data.Dataset, batch_size: int
+    ) -> torch.utils.data.DataLoader:
+        """
+        Create a PyTorch DataLoader.
+
+        Parameters
+        ----------
+        dataset : torch.utils.data.Dataset
+            A PyTorch Dataset.
+
+        Returns
+        -------
+        torch.utils.data.DataLoader
+            A PyTorch DataLoader.
+        """
+        # Calculate new batch size to saturate previous batch size with PSMs
+        pep_per_spec = []
+        for i in range(min(10, len(dataset))):
+            pep_per_spec.append(len(dataset[i][3].split(",")))
+        new_batch_size = max(1, int(batch_size // np.mean(pep_per_spec)))
```

Comment:
I'm a little bit worried about this. If some spectra have significantly more candidates than others, is there a chance of getting an unlucky batch which is far too big and causes memory issues? 

Also, reading through the full dataset to calculate this is not ideal.

Reply:
This does not read through the entire dataset, but rather just the first 10 spectra to get an educated guess as to how many candidate peptides there are going to be for each spectra. I don't see another way to calculate this number other than guessing or explicit user input (or reading through the whole file, which is not ideal).

Ultimately, the method `smart_batch_gen` in `model.py` should ensure that each forward pass has `predict_batch_size` PSMs in it. This calculation is just to make that process slightly faster by trying to have each batch already have approx. `predict_batch_size` PSMs already (so you only need to iterate through a few indexes of the dataloader before doing a forward pass).

Reply:
OK for now. But as you implement the next step, you may not be able to do this because the candidates for each spectrum won't be computed ahead of time anymore.